README
                 -= PolyORB HI (c) ENST 2006 - 2007 =-

--  $Id: README 5244 2007-05-29 13:42:37Z zalila $

Here is a short description of the distribution:

* doc/: documentation,
* examples/: set of examples to test PolyORB HI
* share/: common files
* src/: core of PolyORB HI,
* ChangeLog: release information,
* COPYING: GPLv2 licence document,
* env: environment variables to be set to use PolyORB HI
* README: short description of the distribution.
* NEWS: The new added features
* INSTALL: Installation instructions

For quick build instructions, please refer to the INSTALL file

For more details on this distribution, please refer to doc/polyorb-hi_ug.pdf.



NEWS
New Features of PolyORB-HI:

v1.1 (2008-MM-DD): in addition to v1.0:
	- Message PDU maximal size is deduced automatically from each
	node data types (change 5386).
	
	- The transport high level layer is now generated
	automatically for each node (change 5385).

	- New package hierarchy: PolyORB_HI is the root of all
	non-generated services. PolyORB_HI_Generated is the root of
	all generated services (change 5381 and 5382)
	
	- Added support for AADL "timed" threads: periodic threads
	that are also able to respond sporadically to external events
	(changes 5365, 5366, 5367, 5368).

	- Support of AADL threads that are connected with delayed
	connexions (local case only) (changes 5340 and 5358)

v1.0 (2007-10-01): in addition to v0.9:
	- New example, mode_switch to illustrate the thread mode switch
	(change 5280).
	
	- New example, flight_mgmt to illustrate the new communication
	subsystem (change 5275).
	
	- New communication subsystem based on the AADL standard
	requirements. User code is able now to manipulate thread
	interfaces through several runtime routines. These routines are
	provided by the 'PolyORB_HI.Thread_Interrogators' generic which is
	instantiated once for each thread (changes 5261-5274).
	
	- New demo from SEI CMU, SunSeeker, that demonstrates how to build
	a control system using AADL.
	
	- New demo from IST-ASSERT, MPC, that demonstrates how to build a
	replicated system with multiple instantiations of the same threads
	on different nodes
	
	- New demo to test the import of C functions, import
	
	- PolyORB-HI now uses autoconf/automake

v0.9 (2007-03-07): in addition to v0.8:
	- Renamed project to PolyORB-HI
	
	- Split restriction configuration files to support more
	restrictions for GNAT for HI Systems
	
	- Added support for "pragma Priority" for protected objects
	generated by Ocarina
	
	- All examples are now running, tested on both native and LEON
	targets
	
	- Addition of the SOIF/MTS low level transport
	
	- Simplify configuration by using ADA_PROJECT_PATH variable
	
v0.8 (2006-10-25): in addition to v0.7:
	- Petri net model generation and graphical output from AADL models
	are now integrated,
	
	- the D3.1.3-1 Toy Example can now be fully generated using
	Ocarina,
	
	- updating AADL-to-code transformation rules,
	
	- now automatically launches tsim when the compilation is
	complete, and begins the execution of the compiled node
	
v0.7 (2006-09-29): in addition to v0.6:
	- document the API of PolyORB HI,
	
	- many additions to the POlyORB HI user guide,
	
	- now the 'some_types' example is fully generated by Ocarina,
	
	- added toy example from D3.1.3-1, tuned for Ocarina,
	
	- provided a local version of AADL models for ping, for test with
	LEON targets,
	
	- added many new features in generate_ada.sh script to compile,
	check and run each generated nodes,
	
	- default target for makefile is now context sensitive and either
	process the AADL scenario file, or build the application node,
	
	- added support for schedulability analysis from AADL models using
	Cheddar, see the RMA examples for more details,
	
	- packaging now includes a binary version of ocarina, and a
	command line tool based on cheddar to test for schedulability
		
v0.6 (2006-09-18): in addition to v0.5:
	- added more debug information,
	
	- added "erc32" rule to build with GNAT GAP 2006 for ERC32,
	
	- added coding guidelines to the OPM,
	
	- added release and debug build modes,
	
	- added generic packages to ease code generation from Ocarina,
	
	- corrected AADL 'ping' example for direct code generation from
	Ocarina,
	
	- now the 'ping' example is fully generated by Ocarina,
	
	- added the 'producer_consumer' by M. Heitz,
		
v0.5 (2006-08-02): in addition to v0.4:
	- updated all headers to match current licence, 

	- clean up in the tree superstructure.
	
v0.4 (2006-08-02): in addition to v0.3:
	- remove most occurrences of slices,

	- usual code reformatting, update, etc,
     
	- addition of a documentation.

v0.3 (2006-08-01): in addition to v0.2:
	- new "check" rule for makefiles, allow to check for all rules
	defined in par 7.1 of D3.3.1-2/D3.3.1-2 deliverable,
     
	- provide a low layer wrapper for transport mechanism, allowing
	for multi-platform (sockets on native ones, empty on LEON for
	now),
     
	- added support to build with "GNAT for Leon",
     
	- added a local test that does not depend on transport lower
	layer, works with "GNAT for Leon".

v0.2 (2006-07-28): in addition to v0.1:
	- include code reformatting, cleaning and documentation,
	
	- support for more restrictions (see src/gnat.adc),
	
	- new some_types example.
	
v0.1 (2006-07-26): initial prototype of PolyORB HI with
	-distribution features. This initial release illustrates
	deployment features with a simple ping example.


tools/run_example.sh
#! /bin/sh

# $Id: run_example.sh 5428 2008-10-05 17:44:02Z zalila $

# This script is a shortcut to run a given PolyORB-HI example

# Its argument is the root directory generated for the example

if [ $# -lt 1 ]; then
    echo "Usage $0 <generated_example_directory> [-r]"
    echo "  -r : run the node in reverse (alphabetical) order"
    exit 1;
fi

sample_dir=$1

if test x"$2" = x"-r"; then
    reverse=true;
else
    reverse=false;
fi

if test ! -d ${sample_dir}; then
    echo "${sample_dir} is not a directory";
    exit 2;
fi;

cd ${sample_dir}

# Get the executables

find . -name '*' -perm -u+x -type f \
    -a ! -name '*.o'  -a ! -name '*.ali' \
    -a ! -name '*.ad?' -a ! -name '*.c' -a ! -name '*.h' \
    -a ! -name '*.gpr'  -a ! -name 'Makefile*' > NODES.tmp

if test ${reverse} = "true"; then
    cat NODES.tmp | sort -r > NODES.tmp_2
else
    cat NODES.tmp > NODES.tmp_2
fi

nodes=`cat NODES.tmp_2`
rm NODES.tmp_2 

# Run the nodes

for node in ${nodes}; do
    # Fetch the platform of ${node}

    tmp=`nm ${node} | grep leonbare`;
    
    if test ! x"${tmp}" = x""; then
	target=leon
    else
	tmp=`nm ${node} | grep system__bb__`;

	if test ! x"${tmp}" = x""; then
	    target=erc32
	else
	    target=native
	fi
    fi

    echo "* TARGET = ${target}"
    echo "* running node: ${node}"

    case $target in
	"native")
	    xterm -e "${node}"&
	    sleep 1;;
	"leon")
	    echo "go" | tsim-leon "${node}";;
	"erc32")
	    echo "go" | tsim-erc32 "${node}";;
    esac
done


support/get_sources.sh
#! /bin/bash

# $Id: get_sources.sh 5237 2007-05-25 15:30:35Z zalila $

# Generates the ADA_SPECS_WITH_BODY, ADA_SPECS, ADA_BODIES and
# AADL_SOURCES variables. The AADL_SOURCES variable is generated only
# if there are actually AADL source files

# Has to be executed without parameter in a working directory.

# This script edits the Makefile.am located in the current directory
# and updates the values of the Ada and AADL sources names variables.

# Each Makefile that has to be edited must contain two marker line
# that encapsulate the part of the Makefile modified by this
# script. The first marker line must contain the string: "BEGIN: DO
# NOT DELETE THIS LINE" and the second "END: DO NOT DELETE THIS LINE".

work_dir=`pwd`;

# Find all Ada spec, Ada bodies and AADL source files

ada_specs_tmp=`find . -name "*.ads" | grep -v 'b~'`
ada_bodies_tmp=`find . -name "*.adb" | grep -v 'b~'`
aadl_sources_tmp=`find . -name "*.aadl"`

# Create temporary files

tmp_mk="./Makefile.am.$$"
ada_specs_with_body="./ada_specs_with_body.$$"
ada_specs_without_body="./ada_specs_without_body.$$"
ada_bodies_without_spec="./ada_bodies_without_spec.$$"
aadl_sources="./aadl_sources.$$"

ada_specs_with_body_f="./ada_specs_with_body_f.$$"
ada_specs_without_body_f="./ada_specs_without_body_f.$$"
ada_bodies_without_spec_f="./ada_bodies_without_spec_f.$$"
aadl_sources_f="./aadl_sources_f.$$"

cleanup () {
    rm -f ${tmp_mk}
    rm -f ${ada_specs_with_body}
    rm -f ${ada_specs_without_body}
    rm -f ${ada_bodies_without_spec}
    rm -f ${aadl_sources}
    rm -f ${ada_specs_with_body_f}
    rm -f ${ada_specs_without_body_f}
    rm -f ${ada_bodies_without_spec_f}
    rm -f ${aadl_sources_f}

    exit $1
}

rm -f ${tmp_mk};                    touch ${tmp_mk}

rm -f ${ada_specs_with_body};       touch ${ada_specs_with_body}
rm -f ${ada_specs_without_body};    touch ${ada_specs_without_body}
rm -f ${ada_bodies_without_spec};   touch ${ada_bodies_without_spec}
rm -f ${aadl_sources};              touch ${aadl_sources};

rm -f ${ada_specs_with_body_f};     touch ${ada_specs_with_body_f}
rm -f ${ada_specs_without_body_f};  touch ${ada_specs_without_body_f}
rm -f ${ada_bodies_without_spec_f}; touch ${ada_bodies_without_spec_f}
rm -f ${aadl_sources_f};            touch ${aadl_sources_f}; 


# Fill the different source cathegories

for i in ${ada_specs_tmp}; do
    body="`dirname ${i}`/`basename ${i} .ads`.adb"

    if [ -f ${body} ]; then
	echo "${i}" >> ${ada_specs_with_body}
    else
	echo "${i}" >> ${ada_specs_without_body}
    fi
done

for i in ${ada_bodies_tmp}; do
    spec="`dirname ${i}`/`basename ${i} .adb`.ads"

    if [ ! -f ${spec} ]; then
	echo "${i}" >> ${ada_bodies_without_spec}
    fi
done

for i in ${aadl_sources_tmp}; do
    echo "${i}" >> ${aadl_sources};
done

# Sort the sources alphabetically

ada_specs_with_body_sorted=`cat ${ada_specs_with_body} | sort`
ada_specs_without_body_sorted=`cat ${ada_specs_without_body} | sort`
ada_bodies_without_spec_sorted=`cat ${ada_bodies_without_spec} | sort`
aadl_sources_sorted=`cat ${aadl_sources} | sort`

# ADA_SPECS_WITH_BODY

echo -ne "ADA_SPECS_WITH_BODY =" > ${ada_specs_with_body_f}
for i in ${ada_specs_with_body_sorted} ; do
    echo -ne " \\\\\n	\$(srcdir)/${i}" | sed -e "s,\./,," >> ${ada_specs_with_body_f}
done
echo -e "\n" >> ${ada_specs_with_body_f}

# ADA_SPECS

echo -ne "ADA_SPECS = \$(ADA_SPECS_WITH_BODY)" > ${ada_specs_without_body_f}
for i in ${ada_specs_without_body_sorted} ; do
    echo -ne " \\\\\n	\$(srcdir)/${i}" | sed -e "s,\./,," >> ${ada_specs_without_body_f}
done
echo -e "\n" >> ${ada_specs_without_body_f}

# ADA_BODIES

echo -ne "ADA_BODIES = \$(ADA_SPECS_WITH_BODY:.ads=.adb)" > ${ada_bodies_without_spec_f}
for i in ${ada_bodies_without_spec_sorted} ; do
    echo -ne " \\\\\n	\$(srcdir)/${i}" | sed -e "s,\./,," >> ${ada_bodies_without_spec_f}
done
echo -e "\n" >> ${ada_bodies_without_spec_f}

# AADL_SOURCES

if test ! x"${aadl_sources_sorted}" = x""; then
    echo -ne "AADL_SOURCES =" > ${aadl_sources_f}
    for i in ${aadl_sources_sorted} ; do
	echo -ne " \\\\\n	\$(srcdir)/${i}" | sed -e "s,\./,," >> ${aadl_sources_f}
    done
    echo -e "\n" >> ${aadl_sources_f}
fi

# Edit the Makefile.am file

MAKEFILE="./Makefile.am"

if [ ! -f ${MAKEFILE} ]; then
    echo "${MAKEFILE}: file not found!";
    cleanup 1;
fi;

BEGIN_MSG="BEGIN: DO NOT DELETE THIS LINE"
END_MSG="END: DO NOT DELETE THIS LINE"

LINE_BEGIN=`cat ${MAKEFILE} | grep -n "${BEGIN_MSG}" | awk 'BEGIN{FS=":"}{print $1}'`;
LINE_END=`cat ${MAKEFILE} | grep -n "${END_MSG}" | awk 'BEGIN{FS=":"}{print $1}'`;

# Are all markers present?

if test x"${LINE_BEGIN}" = x"" -o x"${LINE_END}" = x""; then
    echo "${MAKEFILE}: file does not contain markers! => NOT EDITED";
    cleanup 0;
fi;

# Are markers consistent?

if test ${LINE_BEGIN} -gt ${LINE_END}; then
    echo "${MAKEFILE}: Unconsistent markers (END before BEGIN) => NOT EDITED";
    cleanup 1;
fi;

head -${LINE_BEGIN} ${MAKEFILE}  >> ${tmp_mk}
cat ${ada_specs_with_body_f}     >> ${tmp_mk}
cat ${ada_specs_without_body_f}  >> ${tmp_mk}
cat ${ada_bodies_without_spec_f} >> ${tmp_mk}
if test ! x"${aadl_sources_sorted}" = x""; then
    cat ${aadl_sources_f}            >> ${tmp_mk}
fi
tail -n +${LINE_END} ${MAKEFILE} >> ${tmp_mk}

cp -f ${tmp_mk} ${MAKEFILE}
echo "${MAKEFILE} Successfully edited!"

# Cleanup and quit

cleanup 0


support/update_headers.sh
#!/bin/sh

# $Id: update_headers.sh 5393 2008-01-18 12:20:54Z zalila $

# This script has to ve invoked from the PolyORB-HI main directory

headers_po_hi="support/headers_po_hi"

usage () {
    cat <<EOF
Usage: ${0} [OPTIONS]
Options:
        No option:
           Update the Copyright headers only for the newly added files and
           the modified files (using SVN).

        [all]
           Updates the Copyright headers of all files.
EOF
}

if ! test -x ${headers_po_hi}; then
    make -C support headers-po_hi
fi;

if test $# = 0 ; then
    echo "Updating headers for changed and new files"
    
    changed_files=`svn status -q | grep "^[M]" | grep '\.\(\(ad[bs]\)\|\(gpr\)\)' | awk '{print $NF}'`

    added_files=`svn status -q | grep "^[A]" | grep '\.\(\(ad[bs]\)\|\(gpr\)\)' | awk '{print $NF}'`
    
    # For changed files, update the header

    for i in ${changed_files}; do
	${headers_po_hi} ${i}
    done

    # For Added files, create a new header

    for i in ${added_files}; do
	${headers_po_hi} -noh ${i} > /dev/null 2>&1
	${headers_po_hi} ${i}
    done
else
    if test x${1} = xall ; then
	echo "Updating headers for all files"
	
	find . -name "*.adb" -exec ${headers_po_hi} $1 '{}' \;
	find . -name "*.ads" -exec ${headers_po_hi} $1 '{}' \;
    else
	usage 1>&2
	exit 1
    fi
fi

exit 0



support/update_changelog.sh
#!/bin/sh

# $Id: update_changelog.sh 5424 2008-09-24 11:59:12Z zalila $

# This script has to ve invoked from the PolyORB-HI main directory

usage () {
    cat <<EOF
Usage: ./support/`basename ${0}`
      Must be executed from the main directory
EOF
}

if ! test x"${0}" = x"./support/`basename ${0}`"; then
    usage;
    exit 1;
fi;

svn update

latest_rev=`LANG=C head ChangeLog 2>/dev/null\
    | grep '^r' \
    | head -1 \
    | tr ' ' '\n' \
    | head -1 \
    | cut -d'r' -f2`

tmp_file=`mktemp file.XXXXXX`

if test "${latest_rev}" -gt 0; then
    start_rev=`expr ${latest_rev} + 1`

    end_rev=`LANG=C svn info 2>/dev/null \
	| grep "^Revision:" \
	| awk '{print $NF}'`

    if test ${start_rev} -le ${end_rev}; then
	LANG=C svn log -v --incremental -r ${end_rev}:${start_rev} \
	    > ${tmp_file}
	cat ChangeLog >> ${tmp_file}
	mv -f ${tmp_file} ChangeLog
    fi
else
    LANG=C svn log -v > ChangeLog
fi

exit 0



tools/generate_ada.sh
#! /bin/sh

# $Id: generate_ada.sh 5253 2007-06-07 15:04:01Z hugues $

# Wrapper script to exploit and AADL scenario to
# - generate Ada code and build utils (makefiles, project files...),
# - compile the generated code,
# - perform several checks on it
# - run the application (if the user did not request the opposite by
#   mean of the --no-run option)

# This script takes a scenario file which contains the list of the
# AADL files to be handled.

# Syntax of a scenario file :
# - The first line of the scenario file contains the system
#   implementation name.
# - The second line of the scenario contains the target, either native
#   or leon or erc32.
# - The third line of the scenario file contains the list of the
#   application nodes in the order in which they should be launched.
# - Other lines are the user AADL (*.aadl) files used
# - The scenario file has to be in the same directory of the AADL models
#   and the Ada source files.

###############################################################################
# Recall usage of this script

usage () {
   echo "Usage: $0 [options] <scenario_file>";
   echo "  NO_OPTION | --all-stages: Perform all steps and run the"
   echo "    application"
   echo "  --no-run: Perform all steps except the running of the"
   echo "    application"
   echo "  --sched-only: Only perform schedulability analysis"
   echo "  --parse-only: Only parse the AADL model"
   echo "  --help: displays this help message"
   exit 1;
}

###############################################################################
# Scan the command line

scan_command_line () {
    option_set=false
    
    while test $# -gt 0; do
	case "$1" in
	    --all-stages)
		all_stages=true
		option_set=true
		;;
	    --no-run)
		no_run=true
		option_set=true
		;;
	    --sched-only)
		sched_only=true
		option_set=true
		;;
	    --parse-only)
		parse_only=true
		option_set=true
		;;
	    --help | -*)
		usage
		;;
	    *)
		if test x${scenario_file} = x""; then
		    scenario_file=$1;
		else
		    echo "Scenario file already given"
		    usage;
		fi
		;;
	esac
	shift
    done

    if test x${scenario_file} = x""; then
	echo "Scenario file not given"
	usage;
    fi

    if test x${option_set} = x"false"; then
	all_stages=true
    fi
}

###############################################################################
# Parse the scenario

parse_aadl () {
    log_file=parse_$scenario_file.out

    echo
    echo "*** Parsing the AADL description ***"
    echo

    ocarina -e \
	$aadl_files $POLYORB_HI_PATH/share/cheddar/Cheddar_Properties.aadl \
	$POLYORB_HI_PATH/share/ocarina/arao.aadl \
	$target_characteristics > $log_dir/$log_file 2>&1

    a=$?
    if [ $a != 0 ]; then
	echo "error" log in $log_dir/$log_file
	exit $a
    else
	echo [OK] log in $log_dir/$log_file
    fi
}

###############################################################################
# Generate Ada code from the scenario

generate_ada () {
    log_file=aadl_ada_$scenario_file.out

    echo
    echo "*** Generating source code from the AADL description ***"
    echo

    ocarina -g PolyORB-HI-Ada \
	$aadl_files $POLYORB_HI_PATH/share/cheddar/Cheddar_Properties.aadl \
	$target_characteristics > $log_dir/$log_file 2>&1

    a=$?
    if [ $a != 0 ]; then
	echo "error" log in $log_dir/$log_file
	exit $a
    else
	echo [OK] log in $log_dir/$log_file
    fi
}

###############################################################################
# Generate DIA figure from the scenario

generate_dia () {
    log_file=aadl_dia_$scenario_file.out

    echo
    echo "*** Generating EPS figure from the AADL description ***"
    echo

    mkdir $result_dir/fig

    ocarina -p dia \
	$aadl_files $POLYORB_HI_PATH/share/cheddar/Cheddar_Properties.aadl \
	$POLYORB_HI_PATH/share/ocarina/arao.aadl \
	$target_characteristics -o $result_dir/fig/$scenario_file.dia \
	> $log_dir/$log_file 2>&1

    a=$?
    if [ $a != 0 ]; then
	echo "error (ignored)" log in $log_dir/$log_file
    fi

    dia --nosplash -e $result_dir/fig/$scenario_file.eps \
	$result_dir/fig/$scenario_file.dia 2> /dev/null
    a=$?
    if [ $a != 0 ]; then
	echo "error (ignored)" log in $log_dir/$log_file
    else
	echo [OK] log in $log_dir/$log_file
    fi
}


###############################################################################
# Generate Petri nets from the scenario

generate_petri_net () {
    log_file=aadl_pn_$scenario_file.out

    echo
    echo "*** Generating Petriscript from the AADL description ***"
    echo

    ocarina -n \
	$aadl_files $POLYORB_HI_PATH/share/cheddar/Cheddar_Properties.aadl \
	$target_characteristics > $log_dir/$log_file 2>&1

    a=$?
    if [ $a != 0 ]; then
	echo "error" log in $log_dir/$log_file
	exit $a
    else
	echo [OK] log in $log_dir/$log_file
    fi

    # Add files for behavioural specifications

    cp "`ocarina -s`/user_defined.psc" "$result_dir"

    # Add behavioural specifications

    cd $source_dir
    for behave_file in `ls *.psc`; do
        tmp=`echo $behave_file | cut -f1 -d'.'` &&
        echo $work_dir/$result_dir
        cp $behave_file $work_dir/$result_dir/$tmp;
    done;

    mkdir -p $result_dir/pn
    mv partition.psc $result_dir/pn
    cd $result_dir/pn
    pn_include partition.psc &&

    # Calling the PetriScript-to-CAMI wrapper

    new_modular_services -s -fromscratch \
	-petriscript generated_final.psc \
	-store pn.cami >> ../log/$log_file

    rm -f New_modular_services.* tmp_change_storage.* \
	program_processed.petriscript FK_groups

    partition -m pn.cami 2> /dev/null >> ../log/$log_file
    mv model.def pn.def >> ../log/$log_file
    mv model.net pn.net >> ../log/$log_file

    echo
    echo "*** Building SRG of Petri nets ***"
    echo

    log_file=pn_srg_$scenario_file.out

    # Building Symbolic Reachability Graph from model

#    WNSRG pn >> ../log/$log_file
    cd ../..
    echo [OK] log in $log_dir/$log_file

}

###############################################################################
# For each call (#include primitive) to a user defined behavioural spec,
# will build a new instance of behavioural spec, and merge it with main
# generated script.

pn_include () {

    if [ $# -ne 1 ]; then
	echo "usage : $0 main_file"
	exit 1
    fi

    SRC_INC=$1
    TMP_INC=$SRC_INC.tmp

    touch $TMP_INC

    echo "-------------------------------------------------------" >> $TMP_INC
    echo "--                include part                       --" >> $TMP_INC
    echo "--  automatically generated by the 'include' script  --" >> $TMP_INC
    echo "-------------------------------------------------------" >> $TMP_INC

    while read cmd arg1 arg2 next;
      do
      if [ "$cmd" = "#include" ];
	  then
	  if [ -e $arg2 ];
	      then
	      pn_namespace $arg1 $arg2 &&
	      cat $arg1'_'$arg2 >> $TMP_INC
	  fi;
      fi;
    done < $SRC_INC &&

    echo "-------------------------------------------------------" >> $TMP_INC
    echo "--                generated part                     --" >> $TMP_INC
    echo "-------------------------------------------------------" >> $TMP_INC

    echo "" >> $TMP_INC

    cat $SRC_INC >> $TMP_INC

    echo "-------------------------------------------------------" >> $TMP_INC
    echo "--                merging part                       --" >> $TMP_INC
    echo "--  automatically generated by the 'include' script  --" >> $TMP_INC
    echo "-------------------------------------------------------" >> $TMP_INC

    echo "" >> $TMP_INC

    while read cmd arg1 arg2 next;
      do
      if [ "$cmd" = "#include" ];
	  then
	  if [ -e $arg2 ];
	      then
	      rm -f $TMP_INC.tmp &&
	      pn_merge $arg1 $SRC_INC $arg2 $TMP_INC.tmp &&
	      cat $TMP_INC.tmp >> $TMP_INC
	  fi;
      fi;
    done < $SRC_INC &&

    # comment all include calls

    cat $TMP_INC | sed s/"#include"/"-- include"/g > $TMP_INC.tmp
    cat $TMP_INC.tmp > $TMP_INC

    rm -f $TMP_INC.tmp;

    mv $TMP_INC generated_final.psc

}

###############################################################################
# For a given call to a user defined behavioural spec, merge all external
# nodes with corresponding nodes in the generated file.

pn_merge () {

    if [ $# -ne 4 ];
	then
	echo "usage : $0 namespace_str generated_file user_defined_file dest_file"
	exit 1
    fi

    NS_MERGE=$1'_'
    SRC_GEN=$2
    SRC_USR=$3
    TMP_MERGE=$4

    # for each node refered by the User defined file, we
    # put them into the corresponding list

    for var in `grep "append place *" $SRC_USR | cut -f2 -d'"' | cut -f1 -d' ' | sed 's/\("$NS_MERGE"\)\([a-z][a-z0-9_]*\)/\2/'`;
      do
      echo "merge place "'"'"$var"'"'" and place " '"'"$NS_MERGE$var"'";' >> $TMP_MERGE
    done &&

    echo -e "\n" >> $TMP_MERGE

    for var in `grep "append transition *" $SRC_USR | cut -f2 -d'"' | cut -f1 -d' ' | sed 's/\("$NS_MERGE"\)\([a-z][a-z0-9_]*\)/\2/'`;
      do
      echo "merge transition "'"'"$var"'"'" and transition " '"'"$NS_MERGE$var"'";' >> $TMP_MERGE
    done

}

###############################################################################
# Schedulability analysis of the scenario

schedulability_analysis () {
    log_file=scheduling_$scenario_file.out

    echo
    echo "*** Schedulability analysis using Cheddar ***"
    echo

    cheddar_analyzer $aadl_files \
	$POLYORB_HI_PATH/share/ocarina/arao.aadl \
	$POLYORB_HI_PATH/share/ocarina/aadl_project.aadl \
	$POLYORB_HI_PATH/share/ocarina/aadl_properties.aadl \
	$POLYORB_HI_PATH/share/cheddar/Cheddar_Properties.aadl \
	$target_characteristics  > $log_dir/$log_file 2>&1

    a=$?
    if [ $a != 0 ]; then
	echo "error" log in $log_dir/$log_file
    else
	echo [OK] log in $log_dir/$log_file
    fi
}

###############################################################################
# Compiling each node

compile_nodes () {
    echo
    echo "*** Compiling nodes ***"

    OLD_IFS=$IFS
    IFS=' '

    for node in $nodes; do
	log_file=compil_$node.out

	echo
	echo "* compiling node: $node"
	cd $result_dir/$node
	make > ../log/$log_file 2>&1
	a=$?
	if [ $a != 0 ]; then
	    echo "error" log in $log_dir/$log_file
	    exit $a
	else
	    echo [OK] log in $log_dir/$log_file
	fi
	cd ../..
    done
    IFS=$OLD_IFS
}

###############################################################################
# Checking each node

check_nodes () {
    echo
    echo "*** Checking nodes ***"

    OLD_IFS=$IFS
    IFS=' '

    for node in $nodes; do
	log_file=check_$node.out

	echo
	echo "* checking node: $node"
	cd $result_dir/$node
	make check > ../log/$log_file 2>&1
	echo [OK] log in $log_dir/$log_file
	cd ../..
    done
    IFS=$OLD_IFS
}

###############################################################################
# Metrics for each node

metrics_nodes () {
    echo
    echo "*** Computing metrics for each node ***"

    OLD_IFS=$IFS
    IFS=' '

    for node in $nodes; do
	log_file=metrics_$node.out

	echo
	echo "* computing metrics for node: $node"
	cd $result_dir/$node
	gnat metric  > ../log/$log_file 2>&1
	echo [OK] log in $log_dir/$log_file
	cd ../..
    done
    IFS=$OLD_IFS
}

###############################################################################
# Run each node

run_nodes () {
    echo
    echo "*** Running nodes ***"
    echo

    OLD_IFS=$IFS
    IFS=' '

    for node in $nodes; do
	echo "* running node: $node"
	case $target in
	    "native")
		xterm -e "$result_dir/$node/$node"&
		sleep 1;;
	    "leon")
		echo "go" | tsim-leon "$result_dir/$node/$node";;
	    "erc32")
		echo "go" | tsim-erc32 "$result_dir/$node/$node";;
	esac
    done
    IFS=$OLD_IFS
}

###############################################################################
# Parse scenario file

parse_scenario () {
    result_dir="`head -1 $scenario_file`"

    # Remove former generated code

    rm -rf $result_dir
    mkdir $result_dir

    log_dir=$result_dir/log
    mkdir $log_dir

    # Get absolute paths of the source and work directories

    work_dir=`pwd`

    source_dir=`dirname $scenario_file`
    cd $source_dir
    source_dir=`pwd`

    cd $work_dir

    # Get the target

    target=`head -2 $scenario_file | tail -1`

    # Get the target property AADL file

    share_location=$POLYORB_HI_PATH/share/ocarina

    case "$target" in
	"native")
	    target_characteristics=$share_location/native_characteristics.aadl
	    ;;
	"leon"|"erc32")
	    target_characteristics=$share_location/leon_characteristics.aadl
	    ;;
	*)
	    echo "error : unknown traget $target in scenario file"
	    exit 2
	    ;;
    esac

    # Get the list of nodes

    nodes=`head -3 $scenario_file | tail -1`
    n_nodes=` echo $nodes | wc -w`

    # Get the files to be handled

    aadl_files=
    for i in `tail -n +4 $scenario_file | grep aadl$`; do
	aadl_files="$aadl_files $source_dir/$i"
    done
}

###############################################################################
# Main processing begins here

# Parse commande line

all_stages=false
no_run=false
sched_only=false
parse_only=false
scenario_file=""

scan_command_line $@

echo
echo "**********************************************************************"
echo "Ocarina AADL tool collection, (c) ENST"
echo "**********************************************************************"
echo
echo Processing file: $scenario_file

parse_scenario

if test x${all_stages} = x"true" -o x${no_run} = x"true" ; then
    parse_aadl
    #generate_dia
    schedulability_analysis
    #generate_petri_net
    generate_ada
    compile_nodes
    check_nodes
    metrics_nodes
    if test x${no_run} = x"false" ; then
	run_nodes
    fi
elif test x${sched_only} = x"true"; then
    schedulability_analysis
elif test x${parse_only} = x"true"; then
    parse_aadl
fi;

echo
echo "**********************************************************************"
